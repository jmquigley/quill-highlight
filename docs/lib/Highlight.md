<a name="module_Highlight"></a>

## Highlight
A custom Quill highlighting module.  It manages multiple mode instances
that are used to apply text formatting to the contents of the control.

The module is added to the component per the API instructions:
https://quilljs.com/guides/building-a-custom-module/

Once the module is initialized the highlighting mode can be changed with
the `set` function.  This just changes what the reference internally to
another class that handles the formatting.

#### Examples:

```javascript
import {EditorMode, EditorStyle, Highlight} from './modules';

Quill.register('modules/highlight', Highlight);
const quill = new Quill('#editor', {
    formats: '',
    modules: {
        highlight: {
        }
    },
    theme: 'snow'
});
...
const highlight = quill.getModule('highlight');
highlight.set({
    content: 'some value',
    mode: EditorMode.markdown,
    styling: EditorStyle.monokai
});
```

