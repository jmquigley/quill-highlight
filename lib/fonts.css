@font-face {
	font-family: 'Fira Code';
	src: local('FiraCode'), local('FiraCode-Regular'), url('./fonts/FiraCode.ttf');
}

@font-face {
	font-family: 'Inconsolata';
	src: local('Inconsolata'), local('Inconsolata-Regular'), url('./fonts/Inconsolata.ttf');
}

@font-face {
	font-family: 'Source Code Pro';
	src: local('SourceCodePro'), local('SourceCodePro-Regular') url('./fonts/SourceCodePro.ttf');
}
