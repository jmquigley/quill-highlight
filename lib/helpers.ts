// This will get a module level import of a Quill instance, but it will not
// be the one on the window.
//
// globals suck.
//
// import * as q from 'quill';

const debug = require("debug")("quill-markup.helpers");

/**
 * Retrieves the global instance of Quill.  When Quill is used in a browser, it
 * is loaded with the `<script>` tag.  This places the global instance on
 * "global" (or window).  There must only be one copy of this instance.  If you use
 * "require" or import to pull it in it will be scoped to that module and will
 * not be the same as the global version.  Bad things happen when one does this
 * as custom, registered modules and formats will not work unless one registers
 * them on the correct instance that is shown in the browser.   This check and
 * retrieval will ensure that it is uses the global instance.
 * @returns {Quill} the reference to the global Quill instance
 */
export function getQuill() {
	const gQuill = (global as any).Quill;
	if (!gQuill) {
		throw new Error("Quill is not loaded in this application");
	}

	debug("Retrieving global quill instance: %O", gQuill);
	return gQuill;
}

const Quill = getQuill();
const Delta = Quill.import("delta");
const Parchment = Quill.import("parchment");

export {Delta, Parchment, Quill};
