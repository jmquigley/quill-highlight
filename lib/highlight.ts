/**
 * A custom Quill highlighting module.  It manages multiple mode instances
 * that are used to apply text formatting to the contents of the control.
 *
 * The module is added to the component per the API instructions:
 * https://quilljs.com/guides/building-a-custom-module/
 *
 * Once the module is initialized the highlighting mode can be changed with
 * the `set` function.  This just changes what the reference internally to
 * another class that handles the formatting.
 *
 * #### Examples:
 *
 * ```javascript
 * import {EditorMode, EditorStyle, Highlight} from './modules';
 *
 * Quill.register('modules/highlight', Highlight);
 * const quill = new Quill('#editor', {
 *     formats: '',
 *     modules: {
 *         highlight: {
 *         }
 *     },
 *     theme: 'snow'
 * });
 * ...
 * const highlight = quill.getModule('highlight');
 * highlight.set({
 *     content: 'some value',
 *     mode: EditorMode.markdown,
 *     styling: EditorStyle.monokai
 * });
 * ```
 *
 * @module Highlight
 */

'use strict';

import {BaseHighlightMode, Markdown, Text} from './modes';

export enum HighlightMode {
	markdown,
	richedit,
	text
}

export enum HighlightStyle {
	custom = 'custom',
	plain = 'plain',
	monokai = 'monokai'
}

export interface HighlightOptions {
	content?: string;
	custom?: any;
	mode: HighlightMode;
	styling: HighlightStyle;
}

const debug = require('debug')('highlight');

export class Highlight {

	private _modes: Map<HighlightMode, any> = new Map<HighlightMode, any>();
	private _opts: HighlightOptions;
	private _processor: BaseHighlightMode;
	private _quill: any;
	private _styles: Map<string, any> = new Map<string, any>();

	constructor(quill: any, opts: HighlightOptions) {
		debug('Initializing highlight module');

		this._quill = quill;

		this._styles[HighlightStyle.monokai] = require('./styles/monokai.json');
		this._modes[HighlightMode.markdown] = new Markdown(quill);
		this._modes[HighlightMode.text] = new Text(quill);
		this.handleChange = this.handleChange.bind(this);
		quill.on('text-change', this.handleChange);
		this.set(opts);
	}

	public set(opts: HighlightOptions) {

		this._opts = opts = Object.assign({
			mode: HighlightMode.text,
			styling: HighlightStyle.plain,
			custom: {},
			content: ''
		}, opts);

		this._processor = this._modes[opts.mode];
		this._processor.content = opts.content;
		this._processor.style = this._styles[opts.styling];

		debug(`initial content: ${opts.content}`);
		this._quill.setText(opts.content);
	}

	private handleChange(delta: any, old: any, source: string) {
		debug(`highlight change: %o, old: %o, source: %s`, delta, old, source);

		const text: string = this._quill.getText();
		this._processor.highlight(text, 0, text.length);
	}
}
