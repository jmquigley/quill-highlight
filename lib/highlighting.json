{
	"foreground": "black",
	"background": "white",

	"h1": "#ffff00",
	"h2": "#ffd700",
	"h3": "#ffaf00",
	"h4": "#ff8700",
	"h5": "#ff5f00",
	"h6": "#ff0000",

	"admonition": "white",
	"admonitionBackground": "red",
	"attribute": "magenta",
	"blockquote": "gray",
	"bold": "#0087ff",
	"chevron": "gray",
	"comment": "#515a5a",
	"fence": "silver",
	"formula": "#c397d8",
	"hr": "#fc9584",
	"italic": "#af00ff",
	"keywords": "#cc9900",
	"language": "gray",
	"link": "#b9ca4a",
	"linkName": "#7aa6da",
	"linkTitle": "#e7c547",
	"list": "#e2b228",
	"mono": "#8a8a8a",
	"option": "#00ffff",
	"strikethrough": "maroon",
	"underline": "green",
	"wiki": "#7aa6da"
}
