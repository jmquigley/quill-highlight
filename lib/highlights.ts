export const cssHighlights: any = {
	"a11y-dark": "highlights/a11y-dark.css",
	"a11y-light": "highlights/a11y-light.css",
	"agate": "highlights/agate.css",
	"an-old-hope": "highlights/an-old-hope.css",
	"androidstudio": "highlights/androidstudio.css",
	"arduino-light": "highlights/arduino-light.css",
	"arta": "highlights/arta.css",
	"ascetic": "highlights/ascetic.css",
	"atelier-cave-dark": "highlights/atelier-cave-dark.css",
	"atelier-cave-light": "highlights/atelier-cave-light.css",
	"atelier-dune-dark": "highlights/atelier-dune-dark.css",
	"atelier-dune-light": "highlights/atelier-dune-light.css",
	"atelier-estuary-dark": "highlights/atelier-estuary-dark.css",
	"atelier-estuary-light": "highlights/atelier-estuary-light.css",
	"atelier-forest-dark": "highlights/atelier-forest-dark.css",
	"atelier-forest-light": "highlights/atelier-forest-light.css",
	"atelier-heath-dark": "highlights/atelier-heath-dark.css",
	"atelier-heath-light": "highlights/atelier-heath-light.css",
	"atelier-lakeside-dark": "highlights/atelier-lakeside-dark.css",
	"atelier-lakeside-light": "highlights/atelier-lakeside-light.css",
	"atelier-plateau-dark": "highlights/atelier-plateau-dark.css",
	"atelier-plateau-light": "highlights/atelier-plateau-light.css",
	"atelier-savanna-dark": "highlights/atelier-savanna-dark.css",
	"atelier-savanna-light": "highlights/atelier-savanna-light.css",
	"atelier-seaside-dark": "highlights/atelier-seaside-dark.css",
	"atelier-seaside-light": "highlights/atelier-seaside-light.css",
	"atelier-sulphurpool-dark": "highlights/atelier-sulphurpool-dark.css",
	"atelier-sulphurpool-light": "highlights/atelier-sulphurpool-light.css",
	"atom-one-dark-reasonable": "highlights/atom-one-dark-reasonable.css",
	"atom-one-dark": "highlights/atom-one-dark.css",
	"atom-one-light": "highlights/atom-one-light.css",
	"brown-paper": "highlights/brown-paper.css",
	"codepen-embed": "highlights/codepen-embed.css",
	"color-brewer": "highlights/color-brewer.css",
	"darcula": "highlights/darcula.css",
	"dark": "highlights/dark.css",
	"darkula": "highlights/darkula.css",
	"default": "highlights/default.css",
	"docco": "highlights/docco.css",
	"dracula": "highlights/dracula.css",
	"far": "highlights/far.css",
	"foundation": "highlights/foundation.css",
	"github-gist": "highlights/github-gist.css",
	"github": "highlights/github.css",
	"gml": "highlights/gml.css",
	"googlecode": "highlights/googlecode.css",
	"gradient-dark": "highlights/gradient-dark.css",
	"grayscale": "highlights/grayscale.css",
	"gruvbox-dark": "highlights/gruvbox-dark.css",
	"gruvbox-light": "highlights/gruvbox-light.css",
	"hopscotch": "highlights/hopscotch.css",
	"hybrid": "highlights/hybrid.css",
	"idea": "highlights/idea.css",
	"ir-black": "highlights/ir-black.css",
	"isbl-editor-dark": "highlights/isbl-editor-dark.css",
	"isbl-editor-light": "highlights/isbl-editor-light.css",
	"kimbie.dark": "highlights/kimbie.dark.css",
	"kimbie.light": "highlights/kimbie.light.css",
	"lightfair": "highlights/lightfair.css",
	"magula": "highlights/magula.css",
	"mono-blue": "highlights/mono-blue.css",
	"monokai-sublime": "highlights/monokai-sublime.css",
	"monokai": "highlights/monokai.css",
	"night-owl": "highlights/night-owl.css",
	"nord": "highlights/nord.css",
	"obsidian": "highlights/obsidian.css",
	"ocean": "highlights/ocean.css",
	"paraiso-dark": "highlights/paraiso-dark.css",
	"paraiso-light": "highlights/paraiso-light.css",
	"pojoaque": "highlights/pojoaque.css",
	"purebasic": "highlights/purebasic.css",
	"qtcreator_dark": "highlights/qtcreator_dark.css",
	"qtcreator_light": "highlights/qtcreator_light.css",
	"railscasts": "highlights/railscasts.css",
	"rainbow": "highlights/rainbow.css",
	"routeros": "highlights/routeros.css",
	"school-book": "highlights/school-book.css",
	"shades-of-purple": "highlights/shades-of-purple.css",
	"solarized-dark": "highlights/solarized-dark.css",
	"solarized-light": "highlights/solarized-light.css",
	"sunburst": "highlights/sunburst.css",
	"tomorrow-night-blue": "highlights/tomorrow-night-blue.css",
	"tomorrow-night-bright": "highlights/tomorrow-night-bright.css",
	"tomorrow-night-eighties": "highlights/tomorrow-night-eighties.css",
	"tomorrow-night": "highlights/tomorrow-night.css",
	"tomorrow": "highlights/tomorrow.css",
	"vs": "highlights/vs.css",
	"vs2015": "highlights/vs2015.css",
	"xcode": "highlights/xcode.css",
	"xt256": "highlights/xt256.css",
	"zenburn": "highlights/zenburn.css"
};
