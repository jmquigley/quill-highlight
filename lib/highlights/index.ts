require('./agate.style');
require('./androidstudio.style');
require('./arduino-light.style');
require('./arta.style');
require('./ascetic.style');
require('./atelier-cave-dark.style');
require('./atelier-cave-light.style');
require('./atelier-dune-dark.style');
require('./atelier-dune-light.style');
require('./atelier-estuary-dark.style');
require('./atelier-estuary-light.style');
require('./atelier-forest-dark.style');
require('./atelier-forest-light.style');
require('./atelier-heath-dark.style');
require('./atelier-heath-light.style');
require('./atelier-lakeside-dark.style');
require('./atelier-lakeside-light.style');
require('./atelier-plateau-dark.style');
require('./atelier-plateau-light.style');
require('./atelier-savanna-dark.style');
require('./atelier-savanna-light.style');
require('./atelier-seaside-dark.style');
require('./atelier-seaside-light.style');
require('./atelier-sulphurpool-dark.style');
require('./atelier-sulphurpool-light.style');
require('./atom-one-dark.style');
require('./atom-one-light.style');
require('./brown-paper.style');
require('./codepen-embed.style');
require('./color-brewer.style');
require('./darcula.style');
require('./dark.style');
require('./darkula.style');
require('./default.style');
require('./docco.style');
require('./dracula.style');
require('./far.style');
require('./foundation.style');
require('./github-gist.style');
require('./github.style');
require('./googlecode.style');
require('./grayscale.style');
require('./gruvbox-dark.style');
require('./gruvbox-light.style');
require('./hopscotch.style');
require('./hybrid.style');
require('./idea.style');
require('./ir-black.style');
require('./kimbie.dark.style');
require('./kimbie.light.style');
require('./magula.style');
require('./mono-blue.style');
require('./monokai-sublime.style');
require('./monokai.style');
require('./obsidian.style');
require('./ocean.style');
require('./paraiso-dark.style');
require('./paraiso-light.style');
require('./pojoaque.style');
require('./purebasic.style');
require('./qtcreator_dark.style');
require('./qtcreator_light.style');
require('./railscasts.style');
require('./rainbow.style');
require('./routeros.style');
require('./school-book.style');
require('./solarized-dark.style');
require('./solarized-light.style');
require('./sunburst.style');
require('./tomorrow-night-blue.style');
require('./tomorrow-night-bright.style');
require('./tomorrow-night-eighties.style');
require('./tomorrow-night.style');
require('./tomorrow.style');
require('./vs.style');
require('./vs2015.style');
require('./xcode.style');
require('./xt256.style');
require('./zenburn.style');

export const cssHighlights: any = {
	'agate': 'agate.style',
	'androidstudio': 'androidstudio.style',
	'arduino-light': 'arduino-light.style',
	'arta': 'arta.style',
	'ascetic': 'ascetic.style',
	'atelier-cave-dark': 'atelier-cave-dark.style',
	'atelier-cave-light': 'atelier-cave-light.style',
	'atelier-dune-dark': 'atelier-dune-dark.style',
	'atelier-dune-light': 'atelier-dune-light.style',
	'atelier-estuary-dark': 'atelier-estuary-dark.style',
	'atelier-estuary-light': 'atelier-estuary-light.style',
	'atelier-forest-dark': 'atelier-forest-dark.style',
	'atelier-forest-light': 'atelier-forest-light.style',
	'atelier-heath-dark': 'atelier-heath-dark.style',
	'atelier-heath-light': 'atelier-heath-light.style',
	'atelier-lakeside-dark': 'atelier-lakeside-dark.style',
	'atelier-lakeside-light': 'atelier-lakeside-light.style',
	'atelier-plateau-dark': 'atelier-plateau-dark.style',
	'atelier-plateau-light': 'atelier-plateau-light.style',
	'atelier-savanna-dark': 'atelier-savanna-dark.style',
	'atelier-savanna-light': 'atelier-savanna-light.style',
	'atelier-seaside-dark': 'atelier-seaside-dark.style',
	'atelier-seaside-light': 'atelier-seaside-light.style',
	'atelier-sulphurpool-dark': 'atelier-sulphurpool-dark.style',
	'atelier-sulphurpool-light': 'atelier-sulphurpool-light.style',
	'atom-one-dark': 'atom-one-dark.style',
	'atom-one-light': 'atom-one-light.style',
	'brown-paper': 'brown-paper.style',
	'codepen-embed': 'codepen-embed.style',
	'color-brewer': 'color-brewer.style',
	'darcula': 'darcula.style',
	'dark': 'dark.style',
	'darkula': 'darkula.style',
	'default': 'default.style',
	'docco': 'docco.style',
	'dracula': 'dracula.style',
	'far': 'far.style',
	'foundation': 'foundation.style',
	'github-gist': 'github-gist.style',
	'github': 'github.style',
	'googlecode': 'googlecode.style',
	'grayscale': 'grayscale.style',
	'gruvbox-dark': 'gruvbox-dark.style',
	'gruvbox-light': 'gruvbox-light.style',
	'hopscotch': 'hopscotch.style',
	'hybrid': 'hybrid.style',
	'idea': 'idea.style',
	'ir-black': 'ir-black.style',
	'kimbie.dark': 'kimbie.dark.style',
	'kimbie.light': 'kimbie.light.style',
	'magula': 'magula.style',
	'mono-blue': 'mono-blue.style',
	'monokai-sublime': 'monokai-sublime.style',
	'monokai': 'monokai.style',
	'obsidian': 'obsidian.style',
	'ocean': 'ocean.style',
	'paraiso-dark': 'paraiso-dark.style',
	'paraiso-light': 'paraiso-light.style',
	'pojoaque': 'pojoaque.style',
	'purebasic': 'purebasic.style',
	'qtcreator_dark': 'qtcreator_dark.style',
	'qtcreator_light': 'qtcreator_light.style',
	'railscasts': 'railscasts.style',
	'rainbow': 'rainbow.style',
	'routeros': 'routeros.style',
	'school-book': 'school-book.style',
	'solarized-dark': 'solarized-dark.style',
	'solarized-light': 'solarized-light.style',
	'sunburst': 'sunburst.style',
	'tomorrow-night-blue': 'tomorrow-night-blue.style',
	'tomorrow-night-bright': 'tomorrow-night-bright.style',
	'tomorrow-night-eighties': 'tomorrow-night-eighties.style',
	'tomorrow-night': 'tomorrow-night.style',
	'tomorrow': 'tomorrow.style',
	'vs': 'vs.style',
	'vs2015': 'vs2015.style',
	'xcode': 'xcode.style',
	'xt256': 'xt256.style',
	'zenburn': 'zenburn.style'
};
