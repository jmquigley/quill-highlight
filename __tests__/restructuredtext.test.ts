"use strict";

import * as fs from "fs-extra";
import * as path from "path";
// import * as sinon from 'sinon';
import {cleanup, Fixture} from "util.fixture";
import {join} from "util.join";

const debug = require("debug")("restructured.test");

// can't use this before the global require and jsdom initialization
import {Quill} from "../lib/helpers";
let quill: any = null;

import {Markup, MarkupMode} from "../index";

afterAll((done) => {
	cleanup({done, message: path.basename(__filename)});
});

beforeEach(() => {
	const data = fs
		.readFileSync(join(__dirname, "fixtures", "empty-html", "index.html"))
		.toString("utf8");

	document.body.innerHTML = data;
	quill = new Quill("#editor", {
		theme: "snow"
	});

	expect(quill).toBeDefined();
});

test("Create Markup instance with RestructuredText mode", () => {
	const fixture = new Fixture("restructuredtext");
	expect(fixture).toBeDefined();

	const txt = fixture.read("file.rst");
	expect(txt).toBeDefined();

	const markup = new Markup(quill, {
		content: txt,
		mode: MarkupMode.restructuredtext
	});

	expect(markup).toBeDefined();
	expect(markup.quill).toBeDefined();
	expect(markup.editor).toBeDefined();

	markup.refresh();

	const delta = markup.quill.getContents();
	expect(delta).toBeDefined();
	debug("%j", delta);

	expect(delta).toMatchSnapshot();
});
