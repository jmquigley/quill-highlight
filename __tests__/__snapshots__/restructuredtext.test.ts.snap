// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Create Markup instance with RestructuredText mode 1`] = `
Delta {
  "ops": Array [
    Object {
      "attributes": Object {
        "color": "#ffff00",
      },
      "insert": "====",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "#ffff00",
      },
      "insert": "text",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "#ffff00",
      },
      "insert": "====",
    },
    Object {
      "insert": "
lorem ipsum

",
    },
    Object {
      "attributes": Object {
        "color": "#ffd700",
      },
      "insert": "text",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "#ffd700",
      },
      "insert": "====",
    },
    Object {
      "insert": "
lorem ipsum

",
    },
    Object {
      "attributes": Object {
        "color": "#ffaf00",
      },
      "insert": "text",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "#ffaf00",
      },
      "insert": "----",
    },
    Object {
      "insert": "
lorem ipsum

",
    },
    Object {
      "attributes": Object {
        "color": "#ff8700",
      },
      "insert": "text",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "#ff8700",
      },
      "insert": "****",
    },
    Object {
      "insert": "
lorem ipsum

",
    },
    Object {
      "attributes": Object {
        "color": "#ff5f00",
      },
      "insert": "test",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "#ff5f00",
      },
      "insert": "~~~~",
    },
    Object {
      "insert": "
lorem ipsum

",
    },
    Object {
      "attributes": Object {
        "color": "#ff0000",
      },
      "insert": "test",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "#ff0000",
      },
      "insert": "^^^^",
    },
    Object {
      "insert": "
lorem ipsum

",
    },
    Object {
      "attributes": Object {
        "color": "#af00ff",
      },
      "insert": "*italic*",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "#0087ff",
      },
      "insert": "**bold**",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "\`\`",
    },
    Object {
      "attributes": Object {
        "color": "#7aa6da",
      },
      "insert": "monospace",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "\`\`",
    },
    Object {
      "insert": "


",
    },
    Object {
      "attributes": Object {
        "color": "#7aa6da",
      },
      "insert": "reference",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "_",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "\`",
    },
    Object {
      "attributes": Object {
        "color": "#7aa6da",
      },
      "insert": "phrase reference",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "\`_",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "|",
    },
    Object {
      "attributes": Object {
        "color": "#7aa6da",
      },
      "insert": "substitution reference",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "|",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "[",
    },
    Object {
      "attributes": Object {
        "color": "#7aa6da",
      },
      "insert": "1",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "]_",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "[",
    },
    Object {
      "attributes": Object {
        "color": "#7aa6da",
      },
      "insert": "CIT2002",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "]_",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": ".. [",
    },
    Object {
      "attributes": Object {
        "color": "#7aa6da",
      },
      "insert": "*",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "] ",
    },
    Object {
      "attributes": Object {
        "color": "#b9ca4a",
      },
      "insert": "This is the first one. ",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": ".. _",
    },
    Object {
      "attributes": Object {
        "color": "#7aa6da",
      },
      "insert": "Python",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": ": ",
    },
    Object {
      "attributes": Object {
        "color": "#b9ca4a",
      },
      "insert": "http://www.python.org/",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "\`",
    },
    Object {
      "attributes": Object {
        "color": "#7aa6da",
      },
      "insert": "Python ",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "<",
    },
    Object {
      "attributes": Object {
        "color": "#b9ca4a",
      },
      "insert": "http://www.python.org/",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": ">\`_",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": ".. ",
    },
    Object {
      "attributes": Object {
        "color": "#7aa6da",
      },
      "insert": "image",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": ":: ",
    },
    Object {
      "attributes": Object {
        "color": "#b9ca4a",
      },
      "insert": "images/ball1.gif",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": ".. |",
    },
    Object {
      "attributes": Object {
        "color": "#7aa6da",
      },
      "insert": "biohazard| image",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": ":: ",
    },
    Object {
      "attributes": Object {
        "color": "#b9ca4a",
      },
      "insert": "biohazard.png",
    },
    Object {
      "insert": "

",
    },
    Object {
      "attributes": Object {
        "color": "#b9ca4a",
      },
      "insert": "http://www.example.com",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "#b9ca4a",
      },
      "insert": "test@example.com",
    },
    Object {
      "insert": "

",
    },
    Object {
      "attributes": Object {
        "color": "silver",
      },
      "insert": "::",
    },
    Object {
      "insert": "

    let blah: number = \\"stuff\\";",
    },
    Object {
      "attributes": Object {
        "code-block": true,
      },
      "insert": "

",
    },
    Object {
      "insert": "    function test() {",
    },
    Object {
      "attributes": Object {
        "code-block": true,
      },
      "insert": "
",
    },
    Object {
      "insert": "        if (test) {",
    },
    Object {
      "attributes": Object {
        "code-block": true,
      },
      "insert": "
",
    },
    Object {
      "insert": "            for (let i = 0; i < 10; i++) {",
    },
    Object {
      "attributes": Object {
        "code-block": true,
      },
      "insert": "
",
    },
    Object {
      "insert": "                console.log('blah');",
    },
    Object {
      "attributes": Object {
        "code-block": true,
      },
      "insert": "
",
    },
    Object {
      "insert": "            }",
    },
    Object {
      "attributes": Object {
        "code-block": true,
      },
      "insert": "
",
    },
    Object {
      "insert": "        }",
    },
    Object {
      "attributes": Object {
        "code-block": true,
      },
      "insert": "
",
    },
    Object {
      "insert": "    }",
    },
    Object {
      "attributes": Object {
        "code-block": true,
      },
      "insert": "
",
    },
    Object {
      "insert": "

lorem ipsum ",
    },
    Object {
      "attributes": Object {
        "color": "",
      },
      "insert": ":attribute:",
    },
    Object {
      "insert": " and a ",
    },
    Object {
      "attributes": Object {
        "color": "",
      },
      "insert": ":keyword:",
    },
    Object {
      "insert": "

",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "| block text",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "| block text",
    },
    Object {
      "insert": "
",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": "| block text",
    },
    Object {
      "insert": "

",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "* ",
    },
    Object {
      "insert": "lorem ipsum 1
",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "* ",
    },
    Object {
      "insert": "lorem ipsum 2

",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "- ",
    },
    Object {
      "insert": "lorem ipsum 1
",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "- ",
    },
    Object {
      "insert": "lorem ipsum 2

",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "+ ",
    },
    Object {
      "insert": "lorem ipsum 1
",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "+ ",
    },
    Object {
      "insert": "tlorem ipsum 2

",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "#. ",
    },
    Object {
      "insert": "lorem ipsum 1
",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "#. ",
    },
    Object {
      "insert": "lorem ipsum 2

",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "1. ",
    },
    Object {
      "insert": "lorem ipsum 1
",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "2. ",
    },
    Object {
      "insert": "lorem ipsum 2
",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "3. ",
    },
    Object {
      "insert": "lorem ipsum 3
",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "	a. ",
    },
    Object {
      "insert": "lorem ipsum
",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "	b. ",
    },
    Object {
      "insert": "lorem ipsum
",
    },
    Object {
      "attributes": Object {
        "color": "#e2b228",
      },
      "insert": "	c. ",
    },
    Object {
      "insert": "lorem ipsum

",
    },
    Object {
      "attributes": Object {
        "background": "red",
        "color": "white",
      },
      "insert": "TODO",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": ": ",
    },
    Object {
      "insert": "lorem ipsum
",
    },
    Object {
      "attributes": Object {
        "background": "red",
        "color": "white",
      },
      "insert": "FIXME",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": ": ",
    },
    Object {
      "insert": "lorem ipsum
",
    },
    Object {
      "attributes": Object {
        "background": "red",
        "color": "white",
      },
      "insert": "IMPORTANT",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": ": ",
    },
    Object {
      "insert": "lorem ipsum
",
    },
    Object {
      "attributes": Object {
        "background": "red",
        "color": "white",
      },
      "insert": "WARNING",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": ": ",
    },
    Object {
      "insert": "lorem ipsum
",
    },
    Object {
      "attributes": Object {
        "background": "red",
        "color": "white",
      },
      "insert": "TIP",
    },
    Object {
      "attributes": Object {
        "color": "gray",
      },
      "insert": ": ",
    },
    Object {
      "insert": "lorem ipsum

",
    },
  ],
}
`;
